Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java

```diff
@@ -0,0 +1,199 @@
+        this.capacity = capacity;
+        this.dimensions = dimensions;
+        store = new double[capacity * dimensions];
+        refCount = new int[capacity];
+        freeBlockStack = new int[capacity];
+
+        for (int j = 0; j < capacity; j++) {
+            freeBlockStack[j] = capacity - j - 1; // reverse order
+        }
+
+        freeBlockPointer = capacity - 1;
+    }
+
+    /**
+     * @return the number of dimensions in stored points for this PointStore.
+     */
+    public int getDimensions() {
+        return dimensions;
+    }
+
+    /**
+     * @return the maximum number of points that can be stored.
+     */
+    public int getCapacity() {
+        return capacity;
+    }
+
+    /**
+     * @return the reference count for the given index.
+     */
+    public int getRefCount(int index) {
+        return refCount[index];
+    }
+
+    /**
+     * @return the number of points currently stored in this point store.
+     */
+    public int size() {
+        return capacity - freeBlockPointer - 1;
+    }
+
+    /**
+     * Add a point to the point store and return the index of the stored point.
+     *
+     * @return the index value of the stored point.
+     * @throws IllegalArgumentException if the length of the point does not match
+     *                                  the point store's dimensions.
+     * @throws IllegalStateException    if the point store is full.
+     */
+    public int add(double[] point) {
+        checkArgument(point.length == dimensions, "point.length must be equal to dimensions");
+        checkState(freeBlockPointer >= 0, "point store is full");
+
+        int nextIndex = freeBlockStack[freeBlockPointer--];
+
+        for (int j = 0; j < dimensions; j++) {
+            store[j + nextIndex * dimensions] = point[j];
+        }
+
+        refCount[nextIndex] = 1;
+        return nextIndex;
+    }
+
+    /**
+     * Increment the reference count for the given index. This operation assumes
+     * that there is currently a point stored at the given index and will throw an
+     * exception if that's not the case.
+     *
+     * @param index The index value.
+     * @throws IllegalArgumentException if the index value is not valid.
+     * @throws IllegalArgumentException if the current reference count for this
+     *                                  index is nonpositive.
+     */
+    public void incrementRefCount(int index) {
+        checkValidIndex(index);
+        refCount[index]++;
+    }
+
+    /**
+     * Decrement the reference count for the given index.
+     *
+     * @param index The index value.
+     * @throws IllegalArgumentException if the index value is not valid.
+     * @throws IllegalArgumentException if the current reference count for this
+     *                                  index is nonpositive.
+     */
+    public void decrementRefCount(int index) {
+        checkValidIndex(index);
+
+        // TODO it should be the case that if a ref count is > 0 then size should also
```

Comment:
minor. this todo can be deleted.

Reply:
The two `checkState` calls there are sanity checks and should not actually be reachable states. However, I think we should leave them in while we develop this branch to help catch issues early on, and then remove them later once we're confident that our implementation is correct. I've updated the comment to be clearer about this.